libglusterfs/src/client_t.c

```diff
@@ -377,59 +377,33 @@ gf_client_unref(client_t *client)
     }
 }
 
-static int
-__client_ctx_get_int(client_t *client, void *key, void **value)
+static void *
+__client_ctx_get_int(client_t *client, void *key)
 {
     int index = 0;
-    int ret = 0;
 
     for (index = 0; index < GF_CLIENTCTX_INITIAL_SIZE; index++) {
         if (client->scratch_ctx[index].ctx_key == key)
-            break;
-    }
-
-    if (index == GF_CLIENTCTX_INITIAL_SIZE) {
-        ret = -1;
-        goto out;
+            return client->scratch_ctx[index].ctx_value;
     }
 
-    if (value)
-        *value = client->scratch_ctx[index].ctx_value;
-
-out:
-    return ret;
+    return NULL;
 }
 
 static int
 __client_ctx_set_int(client_t *client, void *key, void *value)
 {
     int index = 0;
-    int ret = 0;
-    int set_idx = -1;
 
     for (index = 0; index < GF_CLIENTCTX_INITIAL_SIZE; index++) {
         if (!client->scratch_ctx[index].ctx_key) {
-            if (set_idx == -1)
-                set_idx = index;
-            /* don't break, to check if key already exists
-               further on */
+            client->scratch_ctx[index].ctx_key = key;
```

Comment:
I don't think we can do that. The existing implementation doesn't keep any kind of order. It just stores the data in the first free slot. However, the first free slot for key K1 could be 3, then the slot 2 could be released and a new value is assigned to K1. With the modified code, we won't detect K1 in slot 3 and just use slot 2 for the new value of K1, which is not correct and can have bad consequences later.

Reply:
But we know the key doesn't exist elsewhere, as we've previously searched for it and did not find it? (via the earlier call to __client_ctx_get_int() function) ?

Reply:
I think this detail can be easily overlooked in the future if the callers are reworked or there are new callers to `__client_ctx_set_int()`.

What about something like this:

* `__client_ctx_get_int()` returns the value, if found, but also the address of the ctx that contains the value (if found) or the address of the first free slot (if not found).
* Using this, function `__client_ctx_set_int()` can be greatly simplified (or even removed)
* It can also be used in `__client_ctx_del_int()` to simplify it.